### DISCUSSION TITLE
Improve portability: Usage as library in Go, JS, C/C++/C#, PHP, ...

### DISCUSSION THREAD

**Discussion by binwiederhier** [FiloSottile/age#404] [General]
_(This is a continuation of the [discussion on HN](https://news.ycombinator.com/item?id=30790016).)_

Thank you very much for creating `age`. I'm hoping it'll become the new OpenPGP. It looks fantastic. I have some suggestions below, but of course those are just suggestions. Feel free to ignore them if you like. :smile: 

If it is your desire to grow `age` to be _the_ new file encryption format (and I really hope it is), there should be implementations of it in a wide range of languages, and it should be possible to easily use it in your own language, without having to shell out.

I'm thinking about something like this:

``` go
import ".../age"

func main() {
  age.Encrypt(...)
}
```

and 

``` php
<?php

age_encrypt(...);

```

and so on. You get the idea. 

Given that implementing age in all these languages is a lot of work, I suggest you create a `CONTRIBUTING.md` file and make it clear that you'll looking for people to create/maintain native implementations.

If there is already a way to use age as a library in these languages, then maybe the documentation should be improved to show how to use it like that. I couldn't find anything but the spec and the man page.

Again: Thanks for creating age. And feel free to ignore this. :smile: 

_Side note: I'm very pleased that this discussion has the number 404. :smile:_

**Comment by str4d** [comment]
Current library implementations I am aware of:

- Go: https://github.com/FiloSottile/age (this one 🙂)
- Rust: https://github.com/str4d/rage
- WASM / JS: https://github.com/str4d/wage (wraps `rage`, still in development)
- Python: https://github.com/kushaldas/pyage (wraps `rage`)

**Comment by str4d** [comment]
I'm also looking into using https://github.com/mozilla/uniffi-rs to generate bindings to `rage`, which would enable support for Kotlin, Swift, Python, and Ruby from a common base.

**Comment by binwiederhier** [comment]
Thank you for the responses, and thanks for making rage :smile:. Cool that there are so many projects already. Maybe they should be listed in the main README too, at least the active ones. Though it looks like other than age+rage none of the other ones are active, sadly.
